Use daemon standard conf

tallandtree opened the pull request:
http_proxy support for Upstart systems.
live-restore of the docker daemon is supported.
Default start method for systemd systems is systemd.
Default configuration file for systemd systems is
/etc/docker/daemon.json.

TBD: the bridge-utils package is not yet required, but it will be for one of the next steps: to create the bridge via this role if the bridge is not the default docker0.

tallandtree committed:
Improved Systemd and Upstart configuration.

http_proxy support for Upstart systems.
live-restore of the docker daemon is supported.
Default start method for systemd systems is systemd.
Default configuration file for systemd systems is
/etc/docker/daemon.json.

tallandtree committed:
Fix missing underscore in variable

drybjed commented on defaults/main.yml:
```diff
@@ -161,6 +164,41 @@ docker__listen: [ '{{ docker__tcp_listen }}' ]
+#
+#    docker__debug: true
+#
+docker__debug: false
```
Please use `True` and `False` values for booleans.

drybjed commented on defaults/main.yml:
```diff
@@ -161,6 +164,41 @@ docker__listen: [ '{{ docker__tcp_listen }}' ]
+# .. code-block:: yaml
+#    :linenos:
+#
+#    docker__debug: true
```
You don't need to provide examples like this for each and every variable, they don't add anything important. Variables a few choices could be presented as a simple list, more complicated variables should go into `defaults-detailed.rst` file. But for the Boolean variables like this one you don't need to provide an example `code-block`, since the configuration of a boolean should be obvious.

drybjed commented on defaults/main.yml:
```diff
@@ -161,6 +164,41 @@ docker__listen: [ '{{ docker__tcp_listen }}' ]
+#    docker__graph: /data/docker
+#
+
+docker__graph: []
```
Is there a default `docker__graph` path? if yes, it should be specified in the defaults, unless there are issues with that?

tallandtree replied:
If it is not set, it uses the /var/lib/docker. I'll change it in the defaults.

drybjed commented on tasks/main.yml:
```diff
@@ -60,6 +60,32 @@
+- name: Configure Docker systemd options
+  template:
+    src: 'etc/docker/daemon.j2'
+    dest: '/etc/docker/daemon.json'
```
Template should be named `etc/docker/daemon.json.j2`.

drybjed commented on templates/etc/default/docker.j2:
```diff
@@ -54,5 +54,10 @@ DOCKER_OPTS='{{ docker__tpl_options | join(" ") }}'
+{% if inventory__environment.http_proxy %}
+export http_proxy={{ inventory__environment.http_proxy }}
+export https_proxy={{ inventory__environment.https_proxy }}
+export no_proxy={{ inventory__environment.no_proxy }}
```
These variables should be set in the `debops.docker` default variables so that they can be overridden if necessary, like this:

```
docker__env_http_proxy: '{{ ansible_env.http_proxy | d() }}'
docker__env_https_proxy: '{{ ansible_env.https_proxy | d() }}'
docker__env_no_proxy: '{{ ansible_env.no_proxy | d() }}'
```

And then use the default variables in various configuration files. That way the proxy configuration should be relevant to a paritucular host.

The `inventory__environment` variable can be overriden by `inventory__group_environment` and `inventory__host_environment` variables, so using these directly in the roles isn't optimal. The `debops.environment` role takes care to set these variables in `/etc/environment`, from which other roles should use them via `ansible_env` variable.

drybjed commented on templates/etc/systemd/system/docker.service.j2:
```diff
@@ -2,21 +2,37 @@
 Requires=docker.socket
 
 [Service]
-EnvironmentFile=-/etc/default/docker
```
This will break Docker from Debian configured by this role. You should only remove it for apropriate Docker versions.

tallandtree committed:
Optimized systemd configurations and options added.

Json template improved.
Added debug option to /etc/default/docker template for Upstart systems.

tallandtree committed:
Added docker__env variables for proxy.

And also renamed the json template.

tallandtree committed:
Removed superfluous examples.

tallandtree committed:
Made more robust when changing upstream.

When upstream is changed to true (or false), the old docker version
is removed.
Support for insecure registries and registry mirrors is added.

tallandtree committed:
Removed autoremove and force

Removed docker__insecure_repositories.

tallandtree committed:
Tasks do not use keyword changed.

ypid commented on defaults/main.yml:
```diff
@@ -169,8 +217,8 @@ docker__labels: {}
-#      - '--icc=false'
-#      - '--debug=true'
+#      - "--icc=false"
+#      - "--insecure-registry=10.1.0.0/16"
```
@tallandtree Note that  single quotes are the first level. If you use Jinja, this is where double quotes come into play. We need to write a policy for this … :wink:
